cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,361 @@
+                    )
+                    current_lineages.put(
+                        (
+                            death_waiting_time + lineage["total_time"],
+                            unique_id,
+                            {
+                                "id": unique_id,
+                                "birth_scale": lineage["birth_scale"],
+                                "total_time": death_waiting_time
+                                + lineage["total_time"],
+                                "active": False,
+                            },
+                        )
+                    )
+
+        def node_name_generator() -> Generator[str, None, None]:
+            i = 0
+            while True:
+                yield str(i)
+                i += 1
+        names = node_name_generator()
+
+        # Instantiate the implicit root
+        tree = nx.DiGraph()
+        root = next(names)
+        tree.add_node(root)
+        tree.nodes[root]["birth_scale"] = self.initial_birth_scale
+        tree.nodes[root]["total_time"] = 0
+        current_lineages = PriorityQueue()
+        # Records the nodes that are observed at the end of the experiment
+        observed_nodes = []
+        starting_lineage = {
+            "id": root,
+            "birth_scale": self.initial_birth_scale,
+            "total_time": 0,
+            "active": True,
+        }
+
+        # Sample the waiting time until the first division
+        sample_lineage_event(starting_lineage)
+
+        # Perform the process until there are no active extant lineages left
+        while not current_lineages.empty():
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if self.num_extant:
+                if current_lineages.qsize() == self.num_extant:
+                    remaining_lineages = []
+                    while not current_lineages.empty():
+                        _, _, lineage = current_lineages.get()
+                        remaining_lineages.append(lineage)
+                    min_total_time = remaining_lineages[0]["total_time"]
+                    for lineage in remaining_lineages:
+                        parent = list(tree.predecessors(lineage["id"]))[0]
+                        tree.edges[parent, lineage["id"]]["weight"] += (
+                            min_total_time - lineage["total_time"]
+                        )
+                        tree.nodes[lineage["id"]]["birth_scale"] = tree.nodes[
+                            parent
+                        ]["birth_scale"]
+                        observed_nodes.append(lineage["id"])
+            if not current_lineages.empty():
```

Comment:
I'm going to disagree with @mattjones315  here: I think we should `break` in the above `if self.num_extant:` statement, and remove this `if not current_lineages.empty():` statement altogether, since I find really weird given that we already made the check `if not current_lineages.empty():` in the outer loop.